### ISSUE TITLE
Implement important series-to-series transformers

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#6]
Some series-to-series transformers that would be useful.

### unfitted, single-series, simple

- [ ]  **binning/aggregation transformer**

Behaviour:
returns the sequence of [aggregator application] (e.g., count) in the bins. Index is start time, end time, or index (from start) of bin, depending on index hyper-parameter

Hyper-parameters: 
bin specs - start: time/index, end  : time/index, numbins : integer
index - 'start', 'end', or 'bin'
aggregator - function to apply to values within bin, default = count

alternative to bin specs: index sequence

- [ ] **truncation transformer**

Behaviour:
cuts off any entry in the sequence with index outside [lower, upper]

Hyper-parameters:
 lower, upper : time

- [ ] **simple equal spacing transformer**

Behaviour:
intra/extrapolates series to the nodes by the specified strategy, e.g., fill in nearest or next (careful with boundaries)

Hyper-parameters: 
node specs - start: time/index, end  : time/index, numsteps : integer
index - 'start', 'end', or 'bin'
strategy - 'nearest', 'last' , 'next', 'pw_linear'

alternative to node specs: index sequence

- [ ] **re-indexing transformer**

Behaviour:
changes the index by a the strategy indicated in the reindexing parameter
integer = replace with ascending count
field = get from data frame column

Hyper-parameters: 
strategy - 'integer', 'field'

- [ ] **index extractor transformer**

Behaviour:
creates a series from the index of the series

- [ ] **NA remover transformer**

Behaviour:
removes sequence elements that are numpy.nan

- [ ] **padding transformer**

Behaviour:
pads a sequence/series with value at start or end until it has the desired length

Hyper-parameters: 
where - 'start', 'end'
what - value
length - integer
optional: index treatment

- [ ] **NA imputer**

Behaviour:
Fills in NA values by the specified strategy

Hyper-parameters:
strategy - 'nearest', 'last' , 'next', 'pw_linear'

### unfitted, single-series, reduction

- [ ] **interpolation transformer**

Behaviour:
uses a scikit-learn regressor or classifier to interpolate to the specified index set.
Fits series values against series index, and uses  the regressor/classifier to predict value from index

Hyper-parameters:
index set
estimator - sklearn regressor

- [ ] **Supervised NA imputer**

Behaviour:
Fills in NA values by the specified strategy by using a scikit-learn regressor or classifier. Fits non-NA series values against series index, and uses the regressor/classifier to predict value from index

Hyper-parameters:
strategy - 'nearest', 'last' , 'next', 'pw_linear'

- [ ] advanced: exogeneous or multi-column versions

### unfitted, multiple-series

Note: the below are "unfitted" since they run on the entire series

- [ ] **index homogenization transformer**
Behaviour:
Looks up the indices for all the series and introduces them for all the series. Fills in values at new nodes by the specified strategy.

Hyper-parameters:
strategy - ' NA', 'nearest', 'last', 'next'

### design questions

* would it make sense to create an "interpolator" class which in predict takes a series and an index sequence and returns the values?
* does it make sense to expose dedicated index parameter interfaces in some of the above?

**Comment by jesellier** [comment]
Hello all. I just committed prototype design in DEV for the transformers.

In a nutshell : 
- There is a BASE abstract that store the lambdas functions into a static member dict and contain a single 'transform' method
- Then the different SUBS just needs to override a method that return a dict of parameters to be passed into the lambdas.
- I tried to pass as much functionality as possible into the BASE

Happy to receive feedback/comments !

**Comment by jesellier** [comment]
done A, B, C, F and H

**Comment by TonyBagnall** [comment]
so revised list of incomplete
1) PCA
2) ARMA
3) Cepstrum
4) Cosine

all of which are off the shelf transforms, so should be easy.  Shapelets still need some work so go back on the list
5) Shapelets: Revise and resubmit
and lets add
6) Truncator: Truncate series to the shortest length
7) Padder: zero pad to the longest length
(I started these but got side tracked)

**Comment by mloning** [comment]
Could we allow the `Truncator` to truncate the series to a given length as well (via some kwargs maybe?), that may become useful in forecasting for efficiency reasons (actually used by some methods in M4)

**Comment by fkiraly** [comment]
Updating first post for the 2020 dev sprint.

**Comment by mloning** [comment]
Close this in favour of #483
